`azurerm_eventhub_namespace`: support latest released properties:`local_authentication_enabled`,`public_network_access_enabled`,`minimum_tls_version`

xiaxyi opened the pull request:
1. update eventhub namespace SDK from 2021-01-01-preview to stable 2021-11-01 
2. Add SAS Auth switch for eventhub namespace

katbyte reviewed (approved):
Once minor schema change and this is good to merge 🌵

xiaxyi commented:
Thanks @katbyte updated.

xiaxyi commented:
@katbyte There are two more properties released in the latest api, I added them in this pr.

xiaxyi commented:
related issue: #17483

tombuildsstuff commented on internal/services/eventhub/eventhub_namespace_resource.go:
```diff
@@ -471,6 +512,16 @@ func resourceEventHubNamespaceRead(d *pluginsdk.ResourceData, meta interface{})
 			d.Set("maximum_throughput_units", int(*props.MaximumThroughputUnits))
 			d.Set("zone_redundant", props.ZoneRedundant)
 			d.Set("dedicated_cluster_id", props.ClusterArmId)
+			if props.DisableLocalAuth != nil {
+				localAuthEnable := *props.DisableLocalAuth
+				d.Set("local_authentication_enabled", !localAuthEnable)
+			}
+			if props.PublicNetworkAccess != nil {
+				d.Set("public_network_access_enabled", *props.PublicNetworkAccess == namespaces.PublicNetworkAccessEnabled)
+			}
+			if props.MinimumTlsVersion != nil {
+				d.Set("minimum_tls_version", *props.MinimumTlsVersion)
+			}
```
all 3 of these fields need to be set all of the time, with a default value if they're not returned e.g.

```
publicNetworkAccessEnabled := true
if props.PublicNetworkAccess != nil {
   *props.PublicNetworkAccess == namespaces.PublicNetworkAccessEnabled
}
d.Set("public_network_access_enabled", publicNetworkAccessEnabled)
```

tombuildsstuff commented on internal/services/eventhub/eventhub_namespace_resource.go:
```diff
@@ -190,6 +190,29 @@ func resourceEventHubNamespace() *pluginsdk.Resource {
+			"local_authentication_enabled": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  true,
```
is this also the default state the API returns when provisioning on an older version and upgrading to a newer api version?

xiaxyi replied:
yes, I tried the older version and the default value is the same

tombuildsstuff reviewed (changes requested):
Left a couple of comments inline, we need to ensure there's always a value set for these fields into the state - meaning that we need to ensure we consider what the default value should be when this isn't returned - but if we can fix that up then this should otherwise be good to go 👍

xiaxyi commented:
@tombuildsstuff  Thanks for the comment, I updated the code and added the default value to set back. Feel free to let me know if it looks cool to you.

sebader commented:
I think we need to get some traction on this. The latest version 3.16 already updated the API version. This now seems to have broken the event hub namespace deployment:

This used to work but does no longer:

```
resource "azurerm_eventhub_namespace" "stamp" {

  network_rulesets {
    default_action = "Deny" # Access is only allowed through the private endpoints
  }
```

```
 Error: setting network ruleset properties for Namespace (Subscription: "afffa704-8cbd-46f2-a146-b473f632ecb5"
│ Resource Group Name: "ace2ef6b5-stamp-swedencentral-rg"
│ Namespace Name: "ace2ef6b5-swedencen-evhns"): networkrulesets.NetworkRuleSetsClient#NamespacesCreateOrUpdateNetworkRuleSet: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="BadRequest" Message="This update of networkRuleSets, sets zero IPRule and VirtualNetworkRules, and the DefaultAction is set to Deny. This would render the namespace resources inaccessible for data-plane requests. If the intention was to restrict data-plane requests to private links only, please set the PublicNetworkAccess property to true. Else either set the DefaultAction property to Allow OR add some allowed IPRules or VirtualNetworkRules in the update request. CorrelationId: 64d0d212-c2e9-e1e3-4df4-28dffdd64271"
│ 

```

And if I understand this PR correctly, only once it is merged can `PublicNetworkAccess` be set.

magodo commented:
Tests are passing, except the following 5, which are also failing on the main branch. An investigation on these failures are in progress:

![image](https://user-images.githubusercontent.com/7970698/182785091-17096b46-b09b-451b-84a1-8f1b4b96d141.png)

![image](https://user-images.githubusercontent.com/7970698/182785013-fe94792a-3723-4829-8794-dc8dd107a48f.png)

sebader commented:
thanks @xiaxyi I can confirm this works again now. I replaced
```
network_rulesets {
    default_action = "Deny"
  }
```
with 

     public_network_access_enabled = false
